src/test/e2e.test.ts

```diff
@@ -0,0 +1,106 @@
+import fetch from 'node-fetch';
+import * as dotenv from 'dotenv';
+// import { CACHING_PARAMS } from '../constants';
+// import glob from 'glob';
+// import mime from 'mime';
+import {
+    generateBucketName,
+    buildSite,
+    deploySite,
+    forceDeleteBucket,
+    cleanupExistingBuckets,
+    EnvironmentBoolean,
+    Permission,
+} from './helpers';
+
+jest.setTimeout(120000);
+dotenv.config();
+
+const bucketName = generateBucketName();
+const testingEndpoint = `http://${bucketName}.s3-website-eu-west-1.amazonaws.com`;
+
+console.debug(`Testing using bucket ${bucketName}.`);
+
+beforeAll(async () => {
+    // If a previous test execution failed spectacularly, it's possible the bucket may have been left behind
+    // Here we scan for leftover buckets warn about them/delete them.
+    if (!process.env.SKIP_BUCKET_CLEANUP) {
+        try {
+            await cleanupExistingBuckets(!!process.env.CLEANUP_TEST_BUCKETS);
+        } catch (err) {
+            console.error('[IMPORTANT] Failed to cleanup leftover buckets! All tests will now fail!');
+            console.error(err);
+            throw new Error('[IMPORTANT] Failed to cleanup leftover buckets! All tests will now fail!');
+            // Note that even with this failure, Jest continues running tests but the results are unusable!
```

Comment:
Is calling `process.exit(1)` a possible idea to prevent the tests from continuing?
CI should pick it up as failed because of the non-zero exit code.

Reply:
I definitely tried that, but I don't remember why it wasn't satisfactory. (It was a while ago)

I'll try it again and get back to you.

Reply:
Okay, I checked this and the reason I didn't do this initially is because Jest swallows console.error output and if you call process.exit it never sends that output to the console. But the current solution wasn't much better, since the error message would be displayed near the top of the test output, so it was hard to see at the bottom of the test output why everything had failed.

Solved by using `process.stderr.write` instead.